src/main/scala/edu/twitter/model/impl/neuralnetwork/DataIterator.scala

```diff
@@ -0,0 +1,179 @@
+package edu.twitter.model.impl.neuralnetwork
+
+import java.io.IOException
+import java.util
+import java.util.NoSuchElementException
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+import org.deeplearning4j.models.embeddings.wordvectors.WordVectors
+import org.deeplearning4j.text.tokenization.tokenizer.preprocessor.CommonPreprocessor
+import org.deeplearning4j.text.tokenization.tokenizerfactory.{DefaultTokenizerFactory, TokenizerFactory}
+import org.nd4j.linalg.dataset.DataSet
+import org.nd4j.linalg.dataset.api.DataSetPreProcessor
+import org.nd4j.linalg.dataset.api.iterator.DataSetIterator
+import org.nd4j.linalg.factory.Nd4j
+import org.nd4j.linalg.indexing.{INDArrayIndex, NDArrayIndex}
+
+import scala.collection.JavaConverters._
+
+
+/**
+  * An Iterator used to convert the training (or the testing) data to the format
+  * that the neural network model understands.
+  *
+  * @param data           The labeled tweets
+  * @param wordVectors    WordVectors object
+  * @param batchSize      Size of each minibatch for training
+  * @param truncateLength If reviews exceed
+  */
+@throws[IOException]
+class DataIterator(val data: RDD[Row],
+                   val wordVectors: WordVectors,
+                   val batchSize: Int,
+                   val truncateLength: Int) extends DataSetIterator {
+
+  private final val vectorSize = wordVectors.getWordVector(wordVectors.vocab.wordAtIndex(0)).length
+
+  private val dataList = data.collect()
+  final private val tokenizerFactory: TokenizerFactory = new DefaultTokenizerFactory
+  tokenizerFactory.setTokenPreProcessor(new CommonPreprocessor)
+
+  private var _cursor: Int = 0
+
+  def next(num: Int): DataSet = {
+    if (_cursor >= dataList.length)
+      throw new NoSuchElementException
+    try {
+      nextDataSet(num)
+    } catch {
+      case e: IOException =>
+        throw new RuntimeException(e)
+    }
+  }
+
+  @throws[IOException]
+  private def nextDataSet(num: Int): DataSet = {
+    val reviews = new util.ArrayList[String]()
+    val positive = new util.ArrayList[Boolean]()
+    var i = 0
+
+    while (i < num && _cursor < totalExamples) {
+
+      val msg = dataList(_cursor).getAs[String]("msg")
+      val label = dataList(_cursor).getAs[Double]("label")
+
+      //TODO(elhawaty): Find why some rows return null.
+      if (msg != null) {
+        reviews.add(msg)
+        positive.add(if (label == 1.0) true else false)
+        i += 1
+      } else {
+        println("Invalid tweet:" + dataList(_cursor))
+      }
+
+      _cursor += 1
```

Comment:
I thought this is a circular iteration. if so this would be 
```scala
_cursor = (_cursor + 1) % dataList.size
```

Reply:
No, it has to terminate at some point (after the model has gone through all the data for one time) and we should use reset() whenever we want to reuse the iterator again.